src/nvim/popupmenu.c

```diff
@@ -667,6 +669,114 @@ void pum_redraw(void)
   }
 }
 
+static int pum_position_info_float(win_T *wp, int height,  int max_width)
+{
+  FloatConfig config = FLOAT_CONFIG_INIT;
+  config.relative = kFloatRelativeWindow;
+  config.window = wp->handle;
+  // when pum_above is SW otherwise is NW
+  config.anchor = pum_above ? kFloatAnchorSouth : 0;
+  int col = pum_col + pum_width + pum_scrollbar + 1;
+  // align menu
+  config.row =  pum_row;
+  int right_remain = Columns - col;
+
+  if (right_remain >= 20) {
+    config.width = right_remain;
+    config.col = col - 1;
+  } else if(right_remain < 20 && pum_col >= 20) {
+    config.width = pum_col - 2;
+    config.col = pum_col - config.width -1;
+  } else {
+    return FAIL;
+  }
+
+  config.width = MIN(config.width, max_width);
+  config.height = MIN(wp->w_height_inner, height);
+  config.style = kWinStyleMinimal;
+
+  return win_config_float_preview(curwin, config);
+}
+
+static void pum_preview_settext(char *info, linenr_T *lnum, int *max_width)
```

Comment:
could take `buf_T *buf` param and use `ml_append_buf()` etc functions. (we might not get to avoiding curbuf switching completely in first PR, but we could just as well prepare for it, and avoiding curbuf in new helper functions is good style)

Reply:
yes.. before consider this..then I think it's safe because we are single thread..